yt/frontends/amrvac/data_structures.py

```diff
@@ -242,24 +243,106 @@ def _parse_parameter_file(self):
+        # This is the reason why it uses setdefaultattr: it will only fill in the gaps left
+        # by the "override", instead of overriding them again.
+        # For the same reason, self.units_override is set, as well as corresponding *_unit instance attributes
+        # which may include up to 3 of the following items: length, time, mass, velocity, number_density, temperature
+
+        # note: yt sets hydrogen mass equal to proton mass, amrvac doesn't.
+        mp_cgs = self.quan(1.672621898e-24, 'g')  # This value is taken from AstroPy
+        He_abundance = 0.1  # hardcoded parameter in AMRVAC
+
+        # get self.length_unit if overrides are supplied, otherwise use default
+        length_unit = getattr(self, 'length_unit', self.quan(1, 'cm'))
+
+        # 1. calculations for mass, density, numberdensity
+        if 'mass_unit' in self.units_override:
+            # in this case unit_mass is supplied (and has been set as attribute)
+            mass_unit = self.mass_unit
+            density_unit = mass_unit / length_unit**3
+            numberdensity_unit = density_unit / ((1.0 + 4.0 * He_abundance) * mp_cgs)
+        else:
+            # other case: numberdensity is supplied. Fall back to one (default) if no overrides supplied
+            numberdensity_override = self.units_override.get('numberdensity_unit', (1, 'cm**-3'))
+            numberdensity_unit = self.quan(*numberdensity_override)  # numberdensity is never set as attribute
+            density_unit = (1.0 + 4.0 * He_abundance) * mp_cgs * numberdensity_unit
+            mass_unit = density_unit * length_unit**3
+
+        # 2. calculations for velocity
+        if 'time_unit' in self.units_override:
+            # in this case time was supplied
+            velocity_unit = length_unit / self.time_unit
+        else:
+            # other case: velocity was supplied. Fall back to None if no overrides supplied
+            velocity_unit = getattr(self, 'velocity_unit', None)
+
+        # 3. calculations for pressure and temperature
+        if velocity_unit is None:
+            # velocity and time not given, see if temperature is given. Fall back to one (default) if not
+            temperature_unit = getattr(self, 'temperature_unit', self.quan(1, 'K'))
+            pressure_unit = ((2.0 + 3.0 * He_abundance) * numberdensity_unit * kb_cgs * temperature_unit).in_cgs()
+            velocity_unit = (np.sqrt(pressure_unit / density_unit)).in_cgs()
+        else:
+            # velocity is not zero if either time was given OR velocity was given
+            pressure_unit = (density_unit * velocity_unit ** 2).in_cgs()
+            temperature_unit = (pressure_unit / ((2.0 + 3.0 * He_abundance) * numberdensity_unit * kb_cgs)).in_cgs()
+
+        # 4. calculations for magnetic unit and time
+        time_unit = getattr(self, 'time_unit', length_unit / velocity_unit)  # if time given use it, else calculate
+        magnetic_unit = (np.sqrt(4 * np.pi * pressure_unit)).to('gauss')
```

Comment:
I haven't thoroughly checked any of the math here, leaving that to @neutrinoceros :stuck_out_tongue:

Reply:
I reviewed this a million times already, and I'm not any kind of authority when it comes to magnetic stuff (in amrvac or anywhere). :)
On this particular line, it's very hard to check for it, but I trust @n-claes that he discussed it with AMRVAC's guru, who happens to be his phs advisor

Reply:
But we *did* have a discussion about the unit conversion here and I'm positive that it's correct.

Reply:
Yep this is fine, `velocity_unit` is defined as `sqrt(pressure_unit/density_unit)` so eventually it results in the same normalisation as used by eg. the enzo frontend